### ISSUE TITLE
[Analytics] Setting user ID gets delayed

### ISSUE THREAD

**Issue by dbilgin** [invertase/react-native-firebase#4424]
## Issue

We have been having problems with our data seeing lots of user ids null even though they should be set. Spent a lot of time on why this might be happening and finally I tried the debugger view. I have tried this both on our project and on a newly created react native project. What I did with the code was basically this:

```
await firebaseAnalytics().setUserId('MAAA');
await firebaseAnalytics().setUserProperties({ proptest: 'propvalue' });
await firebaseAnalytics().logEvent('test2');
```

As a result I got the below data:
<img src="https://user-images.githubusercontent.com/15243788/96602911-d08a2780-12f3-11eb-936b-7a6a86963d06.png" width="300" />
You can see that the user id and the property are set **after** the event even though that's not what I did in the code.

Now you may say this is related to [this](https://github.com/invertase/react-native-firebase/issues/4010) issue, but I highly doubt this is just a debugger view issue as we also have problems with our data, most of our users missing the user IDs.

Thinking this might be an expected thing I created a native Android project and added the [Firebase analytics](https://firebase.google.com/docs/analytics/get-started?platform=android) there. I called the functions exactly the same way as I did above and saw this result:

<img src="https://user-images.githubusercontent.com/15243788/96603605-78075a00-12f4-11eb-87d6-e1ff4f5efa35.png" width="300" />

As you can see, here the `test` event is called after the properties are set.

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```
    "@react-native-firebase/analytics": "7.6.8",
    "@react-native-firebase/app": "8.4.6",
    "@react-native-firebase/dynamic-links": "7.5.10",
    "@react-native-firebase/messaging": "7.9.1",
    "@react-native-firebase/perf": "7.4.9",
```

#### `firebase.json` for react-native-firebase v6:

```json
# N/A
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

target 'testapp' do
  config = use_native_modules!
  use_react_native!(:path => config["reactNativePath"])

  target 'testappTests' do
    inherit! :complete
    # Pods for testing
  end

  # Enables Flipper.
  #
  # Note that if you have use_frameworks! enabled, Flipper will not work and
  # you should disable these next few lines.
  use_flipper!
  post_install do |installer|
    flipper_post_install(installer)
  end
end

```

#### `AppDelegate.m`:

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [x] my application is an AndroidX application?
- [x] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [x] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
System:
    OS: macOS 10.15.7
    CPU: (8) x64 Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
    Memory: 3.28 GB / 16.00 GB
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 12.13.1 - ~/.nvm/versions/node/v12.13.1/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.12.1 - ~/.nvm/versions/node/v12.13.1/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.9.3 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: iOS 14.0, DriverKit 19.0, macOS 10.15, tvOS 14.0, watchOS 7.0
    Android SDK:
      API Levels: 23, 28, 29, 30
      Build Tools: 28.0.3, 29.0.2, 30.0.2
      System Images: android-30 | Google Play Intel x86 Atom
      Android NDK: Not Found
  IDEs:
    Android Studio: 4.0 AI-193.6911.18.40.6626763
    Xcode: 12.0.1/12A7300 - /usr/bin/xcodebuild
  Languages:
    Java: 14.0.2 - /usr/bin/javac
    Python: 2.7.16 - /Users/user/.pyenv/shims/python
  npmPackages:
    @react-native-community/cli: Not Found
    react: 16.13.1 => 16.13.1
    react-native: 0.63.3 => 0.63.3
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [x] Both
- **`react-native-firebase` version you're using that has this issue:**
    "@react-native-firebase/analytics": "7.6.8",
    "@react-native-firebase/app": "8.4.6",
    "@react-native-firebase/dynamic-links": "7.5.10",
    "@react-native-firebase/messaging": "7.9.1",
    "@react-native-firebase/perf": "7.4.9",
- **`Firebase` module(s) you're using that has the issue:**
  - `e.g. Instance ID`
- **Are you using `TypeScript`?**
  - Yes - 4.0.3

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by mikehardy** [comment]
Very diligent @dbilgin checking the native quickstart, that would have been my first question.

Second question is if this is a race condition - do you have a sample size where you're reasonably sure that native quickstart test works *every* time, and RNFB is failing some/most of the time? (does RNFB work ever?)

Last question then is basically the "big what the heck?" - have you tried instrumenting where RNFB calls the native APIs to see when they are called (are they out of order? if so, why? etc) - this instrumentation in the native code is where I would head next

**Comment by dbilgin** [comment]
After chasing this for a long time I've found some closure finally and wanted to update it here. Thank you for your comment @mikehardy, it did direct me towards the correct direction.

Reaching out to firebase and doing more tests, I found out that even if we set the user id as soon as the firebase analytics is activated, there is no way to add the user id to events such as `first_open`. That was the first discovery

The second important part was that actually the debugger view is not very dependable. They've told me from firebase that the order of the events on this debugger view can be completely wrong, which makes me wonder what the debugger view is for then, but that's a different subject. I was able to follow what's going on through [verbose logs](https://firebase.google.com/docs/analytics/events?platform=android#view_events_in_the_android_studio_debug_log) and that was the solution about events like `first_open`.

We were enabling the firebase analytics according to GDPR compliance and there I could see that I can never set the user id before the first automatic events are logged.

With this I think I can close this issue.
